### ISSUE TITLE
MSCrashes not triggering, even though there was a crash in the previous session

### ISSUE THREAD

**Issue by lwouis** [microsoft/appcenter-sdk-apple#2146]
### **Description**

Some users have reported the app crashing at launch, and not getting prompted to send a crash report on further attempts. After investigating, I was able to reproduce that scenario locally. Indeed, I can start the app, have MSCrashes correctly initialized. Evidence being that:

```swift
debugPrint(MSCrashes.isEnabled(), MSCrashes.hasCrashedInLastSession(), MSCrashes.lastSessionCrashReport())
```

prints

```
 true true Optional(<MSErrorReport: 0x7f9efd571170>)
```

Yet, when I add a delegate which implements this, it prints nothing:

```swift
func crashes(_ crashes: MSCrashes!, shouldProcessErrorReport errorReport: MSErrorReport!) -> Bool {
    debugPrint("hey")
    return true
}
```

Here are the logs with `MSAppCenter.setLogLevel(MSLogLevel.verbose)`:

```
2020-08-11 16:23:45.991 AltTab[36212:1141681] [AppCenter] VERBOSE: -[MSDelegateForwarder addTraceBlock:]_block_invoke_2/88 Start buffering traces.
2020-08-11 16:23:45.992 AltTab[36212:1141681] [AppCenter] DEBUG: -[MSDelegateForwarder setEnabledFromPlistForKey:]_block_invoke/278 Delegate forwarder for info.plist key 'AppCenterAppDelegateForwarderEnabled' enabled. It may use swizzling.
2020-08-11 16:23:45.992 AltTab[36212:1141681] [AppCenter] VERBOSE: +[MSDelegateForwarder flushTraceBuffer]/106 Stop buffering traces, flushed.
2020-08-11 16:23:45.992 AltTab[36212:1141681] [AppCenter] DEBUG: -[MSSessionContext init]/42 10 session(s) in the history.
2020-08-11 16:23:45.993 AltTab[36212:1141681] [AppCenter] VERBOSE: -[MSSessionContext setSessionId:]/66 Stored new session with id:(null) and timestamp: 2020-08-11 07:23:45 +0000.
2020-08-11 16:23:45.994 AltTab[36212:1141681] [AppCenter] DEBUG: -[MSUserIdContext init]/54 10 userId(s) in the history.
2020-08-11 16:23:45.997 AltTab[36212:1141681] [AppCenter] DEBUG: -[MSDBStorage initWithSchema:version:filename:]_block_invoke/31 SQLite global configuration successfully updated.
2020-08-11 16:23:46.001 AltTab[36212:1141681] [AppCenter] INFO: -[MSAppCenter configureWithAppSecret:transmissionTargetToken:fromApplication:]/317 App Center SDK configured successfully.
2020-08-11 16:23:46.001 AltTab[36212:1141681] [AppCenter] VERBOSE: -[MSAppCenter startServices:withAppSecret:transmissionTargetToken:fromApplication:]/343 Start services MSCrashes from an application
2020-08-11 16:23:46.005 AltTab[36212:1141681] [AppCenterCrashes] DEBUG: -[MSCrashes configureCrashReporterWithUncaughtExceptionHandlerEnabled:]/665 EnableUncaughtExceptionHandler is set to YES
2020-08-11 16:23:46.005 AltTab[36212:1141681] [AppCenterCrashes] VERBOSE: -[MSCrashes configureCrashReporterWithUncaughtExceptionHandlerEnabled:]/675 Enabled Mach exception handler.
2020-08-11 16:23:46.010 AltTab[36212:1141681] [AppCenterCrashes] DEBUG: -[MSCrashes configureCrashReporterWithUncaughtExceptionHandlerEnabled:]/707 Exception handler successfully initialized.
2020-08-11 16:23:46.010 AltTab[36212:1141681] [AppCenterCrashes] DEBUG: +[MSApplicationForwarder registerForwarding]/195 Application forwarder for info.plist key 'AppCenterApplicationForwarderEnabled' disabled. It won't use swizzling.
2020-08-11 16:23:46.010 AltTab[36212:1141681] [AppCenterCrashes] DEBUG: -[MSCrashes applyEnabledState:]/389 The application crashed in the last session.
2020-08-11 16:23:46.022 AltTab[36212:1141681] [AppCenterCrashes] INFO: -[MSCrashes applyEnabledState:]/412 Crashes service has been enabled.
2020-08-11 16:23:46.027 AltTab[36212:1141681] [AppCenterCrashes] VERBOSE: -[MSCrashes startWithChannelGroup:appSecret:transmissionTargetToken:fromApplication:]/457 Started crash service.
2020-08-11 16:23:46.027 AltTab[36212:1141681] [AppCenterCrashes] VERBOSE: -[MSCrashes channel:didPrepareLog:internalId:flags:]/540 Storing a log to Crashes Buffer: (sid: (null), type: startService)
2020-08-11 16:23:46.027 AltTab[36212:1141681] [AppCenterCrashes] VERBOSE: -[MSCrashes channel:didPrepareLog:internalId:flags:]/551 Found an empty buffer position.
2020-08-11 16:23:46.027 AltTab[36212:1141722] [AppCenter] DEBUG: -[MSChannelUnitDefault enqueueItem:flags:]_block_invoke/154 Saving log, type: startService, flags: 1.
2020-08-11 16:23:46.029 AltTab[36212:1141722] [AppCenter] VERBOSE: -[MSLogDBStorage saveLog:withGroupId:flags:]_block_invoke/129 Log is stored with id: '905'
2020-08-11 16:23:46.029 AltTab[36212:1141722] [AppCenterCrashes] VERBOSE: -[MSCrashes channel:didCompleteEnqueueingLog:internalId:]/593 Deleting a log from buffer with id 8B511CE5-8011-43B8-AA94-539DF55BA812
```

Note that the crash happens on the main thread, trying to unwrap an optional that's `nil`:

```
Fatal error: Unexpectedly found nil while unwrapping an Optional value: file /Users/louis.pontoise/git_clones/alt-tab-macos/src/ui/preferences-window/tabs/ControlsTab.swift, line 110

Process finished with exit code 132 (interrupted by signal 4: SIGILL)
```

### **Details**

1. Which SDK version are you using?
    - `pod 'AppCenter/Crashes', '3.3.1'`
2. Which OS version did you experience the issue on?
    - macOS 10.15.4, 10.15.5, 10.15.6, 10.16.0
3. Which Cocoapods version are you using (run `pod --version`)?
    - 1.9.1
4. What device version did you see this error on?  Were you using an emulator or a physical device?
    - 4 or 5 different models of macbook pro and air
5. What language are you using?
    - Swift
6. What third party libraries are you using?
    - LetsMove, Sparkle, ShortcutRecorder, SwiftyMarkdown, Preferences

**Comment by MatkovIvan** [comment]
Hi again 👋 

> the app crashing at launch

This is why. I watched the attached video in the original thread - the app crashes each time, so the SDK doesn't have the time to show or report anything - the reporter in this case is "**in-process**" so it lives only with your app, we do not spawn any background process or so.
The collected crashes will be sent once the app starts normally.

PS this is a first line in our [troubleshooting doc](https://docs.microsoft.com/en-us/appcenter/diagnostics/troubleshooting):

> If your app crashes on startup, the SDK can't send any crash reports. This limitation applies to any third-party crash reporting service.

**Comment by lwouis** [comment]
I'm a bit confused by the logic. This is how I imagine the crash reporter logic:

* App init
* `MSAppCenter.start(secret, withServices: [MSCrashes.self])`
* `MSCrashes` checks for a crash in the previous session. If found, it triggers the userConfirmationHandler set with `MSCrashes.setUserConfirmationHandler`.

All this being on the main thread means it would spawn the crash report UI and report the crash before the main thread reaches the crash point.

What is `MSCrashes` waiting for, before it calls the handler?

**Comment by MatkovIvan** [comment]
Yes, it triggers on the main thread after start, but not immediately (only after 1 second). It allows to handle some corner cases and keep compatibility with "wrappers" (Xamarin, React-Native, etc.) that have custom initialization logic. Since the network request is async anyway, it doesn't really matter that UI was shown or not.

**Comment by lwouis** [comment]
The scenario where the app crash on launch is precisely the one that motivated the integration of AppCenter SDK. Indeed, in this scenario, the user can't use the feedback form built into the app to report the problem, as the app simply won't launch.

I must say, tying the crash reporter lifecycle to the app lifecycle seems counter-intuitive to me. I feel a crash reporter should be external, so that it can observe and report about the app crashing without being threatened to be terminated at any time.

In the current in-process implementation, I think the use-case of init crashes could be supported by having an option to have the AppCenter SDK report the crash be sync instead of async. This way, users can guarantee that the crash report phase is done before moving on to trying to launch the app, which can always fail.

Of course, I don't know if this suggestion makes sense. That's why I would like to rely on your expertise.

How can I cover this early crash use-case with your SDK?

**Comment by MatkovIvan** [comment]
Yes, in-process crash reporting has some limitations.

> I feel a crash reporter should be external

On iOS/tvOS/watchOS you can't run it outside of the process. On macOS it's technically possible, but in this case the crash reporter doesn't have access to the state of crashed app.

The sending operation relies on network and can be postponed to the next connect/launch. This is not the thing that an app should wait before start.
The only way to report early crashes (without subsequent successful launches) is to rely on [system crash reporter](https://developer.apple.com/library/archive/technotes/tn2151/_index.html).

**Comment by lwouis** [comment]
> The sending operation relies on network and can be postponed to the next connect/launch. This is not the thing that an app should wait before start.

Why not? If the app is known to have crashed, I think it's reasonable to have the crash reporting be a first-class process as part of the UX of the app. Many huge apps (e.g. Microsoft Outlook) use this approach. What about providing this as an option for users who desire this UX with its trade-offs?

> The only way to report early crashes (without subsequent successful launches) is to rely on system crash reporter

Actually that's something I would love to understand: what's the difference between AppCenter SDK crash report and a System crash report? Some users send me reports manually from the Console.app > Crash Report section, and they look very similar to what I find in AppCenter. Is AppCenter SDK simply sending these files to the Portal?

Also, to clarify your answer a bit; you're suggesting that I implement grabbing these system reports from their folders and upload them to AppCenter Portal using the HTTPS API?

**Comment by MatkovIvan** [comment]
> What about providing this as an option for users who desire this UX with its trade-offs?

Currently, it's not supported. But feel free to file separate feature request to our [App Center GitHub repo](https://github.com/microsoft/appcenter/issues).

> what's the difference between AppCenter SDK crash report and a System crash report?

"apple like" formatting is provided by portal, SDK doesn't use this text representation.

> Is AppCenter SDK simply sending these files to the Portal?

No. App Center uses [PLCrashReporter](https://github.com/microsoft/plcrashreporter) (low level API) under the hood to collect state of the crashed process.

> you're suggesting that I implement grabbing these system reports from their folders and upload them to AppCenter Portal using the HTTPS API?

No, uploading this format is not supported. I meant that Apple App Store provides access to app's reports collected via system crash reporter.

**Comment by lwouis** [comment]
> feel free to file separate feature request to our App Center GitHub repo.

[Done](https://github.com/microsoft/appcenter/issues/1954) 👍

> App Center uses PLCrashReporter (low level API) under the hood to collect state of the crashed process

Could you tell me what I would lose if I collected system crash reports in a backend, instead of using AppCenter portal? Reports from both seem very similar to me. What is the added-value from PLCrashReporter compared to the system crash reporter? What extra info, or better layout does PLCrashReporter provide?

> I meant that Apple App Store provides access to app's reports collected via system crash reporter.

[My app](https://github.com/lwouis/alt-tab-macos) is not on the AppStore, so I would need to handle crash reporting myself.

**Comment by MatkovIvan** [comment]
> what I would lose

Symbolycation, grouping, UI for that, [attachments](https://docs.microsoft.com/en-us/appcenter/sdk/crashes/android#add-attachments-to-a-crash-report), breadcrumbs (session matching with Analytics, the features from ecosystem and integration with other services.

I'd consider this crash as a corner case (documented one). As you probably already saw, in-process crash reporter covers a lot of cases.

**Comment by MatkovIvan** [comment]
Ah, also there is a comparison of the reporters (it's a bit outdate, App Center ~ HockeyApp there): https://www.crashprobe.com/mac/
It compares just details in the generated reports

**Comment by MatkovIvan** [comment]
Closing this issue as it looks like resolved. Feel free to comment/reopen if you have more questions
